File: Python3/343. Integer Break.py
class Solution:
    def integerBreak(self, n):
        """
        :type n: int
        :rtype: int
        """
        ans = [1] * (n+1)
        for i in range(2, n+1):
            ans[i] = ans[i-1]
            for j in range(1, 10):
                if i-j > 0 and ans[i-j] * j > ans[i]:
                    ans[i] = ans[i-j] * j
                if i < n: # 关键是这一步, 如果不等于和, 那么可以不拆分
                    ans[i] = max(ans[i], i)
        return ans[n]

File: Golang/343. Integer Break.go
func integerBreak(n int) int {
    ans := make([]int, n+1)
    ans[1] = 1
    for i := 2; i <= n; i++ {
        ans[i] = ans[i-1]
        for j := 1; j <= 9; j++ {
            if i-j > 0 && ans[i-j]*j > ans[i] {
                ans[i] = ans[i-j] * j
            }
            if i < n && ans[i] < i {
                ans[i] = i
            }
        }
    }
    return ans[n]
}
